Refactor async operations on individual package results

lennym opened the pull request:
Previously the callback for the lambda was fired before the individual results had been fully saved to s3 and emitted to SNS. This meant that results were lost from the end of the set and never returned to the client.

By refactoring to move the sending of individual results to inside the iteration over hotel ids then we can be sure that the lambda does not terminate until all results have successfully been returned to the client.

lennym committed:
Refactor async operations on individual package results

Previously the callback for the lambda was fired before the individual results had been fully saved to s3 and emitted to SNS. This meant that results were lost from the end of the set and never returned to the client.

By refactoring to move the sending of individual results to inside the iteration over hotel ids then we can be sure that the lambda does not terminate until all results have successfully been returned to the client.

Kumjami commented on index.js:
```diff
@@ -1,77 +1,72 @@
+      return searchForPackages(params);
+    })
+    .then((results) => {
+      return results.filter(r => r);
```
what does this line do? 

``` js
    return results.filter(r => r);
```

lennym replied:
Strips the empty entries from the results.

If a hotel has no availability then it will create an empty entry in the array. We're not interested in keeping these in the results so filter them out.

Kumjami commented on package.json:
```diff
@@ -34,9 +34,10 @@
   "dependencies": {
-    "async": "2.0.0-rc.5",
     "aws-lambda-helper": "^2.18.0",
+    "bluebird": "^3.4.1",
```
is it worth it to include a promise library? 
can native 4.3.2 node version do the same without adding a dependency?

lennym replied:
Yes. Node natively gives you `new Promise((resolve, reject) => {...})` but that is all. The library gives us things like iterators, which we use heavily.

lennym commented:
Worth pointing out that this has been deployed as `lambda-ne-dynamic-package-provider-v2` and is listening to the search request SNS events, so if you're looking on an environment that uses the `ci` pipeline then you should already be seeing results from this provider.

lennym committed:
Refactor tests to hit mocked apis instead of real api

Use nock to mock http api endpoints and assert that calls are made as expected.

lennym committed:
Add a test env file for the unit tests

This removes the need for a user to configure their own environment variables to be able to run the tests. Since all communication withthird party services is mocked there is no need for these values to be "real".

lennym commented:
@Kumjami This is good to go now. Sorry for the massive diff. :-(

You can largely disregard any changes in ./test/fixtures though, which amount to most of it.

lennym committed:
Correctly load the provider reference

This is used to id the results on the client. This is not a good idea, and should be fixed in the client, but for now it is necessary to load results correctly.

lennym committed:
Update helper version

This will add provider names to the results sent back to the client, which will help with debugging.

nelsonic commented:
Wow! You weren't kidding when you said "Big PR"... 😮 
is anything else higher priority or should I get started reviewing this now?

lennym committed:
Send searchComplete event to client

Emits an event to the client to mark the search for this provider as complete.

lennym commented:
@nelsonic This is causing a real bug in prod, so should be fairly high priority.

For what it's worth, _most_ of the diff is in test fixtures, which can be fairly safely disregarded (they're just copies of the results from the real APIs).

lennym commented:
Also, the new code itself isn't that big, so you may find it easier just to review the new code, rather than the changes.

nelsonic commented on lib/api_request.js:
```diff
@@ -46,22 +46,18 @@ function sort_by_price_asc (a, b) { // as a user I want to see cheapest holidays
+  http.request(options, function (err, data) {
+    /* istanbul ignore if */
+    if (err) {
+      AwsHelper.log.info({ err: err }, 'Error retrieving hotel info from API');
```
does calling back with an error here display an error message in UI?

lennym replied:
No. Errors in search providers never make it back to the UI. 

We can add some error feedback to the UI, but I've been under the impression that this is not desired, and that the reasoning behind multiple providers is to be fault tolerant, and that if one provider errors we just show the results from other providers.

nelsonic replied:
Yeah, with multiple providers the UI should display _some_ result, 
we can re-assess if this is a requirement if we get a zero-results situation. 👍

nelsonic commented on lib/api_request.js:
```diff
@@ -74,70 +70,30 @@ function get_hotel_info (hid, callback) {
+        /* istanbul ignore if */
+        if (err) {
+          AwsHelper.log.warn({ err: err, hotelId: params.hotelId, request: options }, 'Unable to get hotel info');
+          callback();
```
should we `return callback(err);` here?

lennym replied:
Well... we don't want errors in any one particular hotel's api request pipeline to cause errors in the full result set (fault tolerance again), so we certainly need to ignore this error at some level. (_Note - [it's ignored in the current code](https://github.com/numo-labs/lambda-ne-dynamic-package-provider/blob/814cb4d38645b48eb27e3b93c5c9d11246ff3006/lib/api_request.js#L113)_)

There is an argument to be made that this module as a standalone component should throw errors as usual though, and that they should be handled at the call site. If I were building from scratch I'd probably do that, it's just as a progression from the previous iteration that it's the way it is.

tl;dr - probably yes, I'll update it.

_Edit - [updated](https://github.com/numo-labs/lambda-ne-dynamic-package-provider/pull/13/commits/18adeebda10583903b8ccdd3221fcf7546fbd96d)_

lennym committed:
Allow api interator to throw errors

It's better module design to allow the module to throw, and then swallow the error at the call site as it allows the child module to be re-used in different contexts.

nelsonic commented on lib/http_request.js:
```diff
@@ -9,26 +8,25 @@ var AwsHelper = require('aws-lambda-helper');
-      var err = null;
-      try { // avoid fatal error if ONE of the http requests has invalid JSON
-        json = JSON.parse(resStr);
+    res.pipe(bl((err, body) => {
```
@lennym would you mind adding a comment clarifying what [`bl`](https://www.npmjs.com/package/bl) is doing here? (_thanks!_)

nelsonic commented:
While I'm not a "_fan_" of using Promises, this use-case is good and the code appears to solve the issue.
merging. 👍

lennym commented:
@nelsonic I really recommend becoming a Promise fan. Constructing complex async workflows is far easier, code is more readable and far less repetitive, and error handling is orders of magnitude easier.
